checks.d/powerdns_recursor.py

```
@@ -0,0 +1,79 @@
+        'cache-misses',
+        'noerror-answers',
+        'outgoing-timeouts',
+        'questions',
+        'servfail-answers',
+        'tcp-outqueries',
+        'tcp-questions',
+    ]
+
+    SERVICE_CHECK_NAME = 'powerdns.recursor.can_connect'
+
+    def check(self, instance):
+        config, tags = self._get_config(instance)
+        stats = self._get_pdns_stats(config)
+        for stat in stats:
+            if stat['name'] in PowerDNSRecursorCheck.GAUGE_METRICS:
+                self.gauge('powerdns.recursor.{}'.format(stat['name']), float(stat['value']), tags=tags)
+            elif stat['name'] in PowerDNSRecursorCheck.RATE_METRICS:
+                self.rate('powerdns.recursor.{}'.format(stat['name']), float(stat['value']), tags=tags)
+
+    def _get_config(self, instance):
+        required = ['host', 'port', 'api_key']
+        for param in required:
+            if not instance.get(param):
+                raise Exception("powerdns_recursor instance missing %s. Skipping." % (param))
+
+        host = instance.get('host')
+        port = int(instance.get('port'))
+        api_key = instance.get('api_key')
+        tags = instance.get('tags', [])
+
+        Config = namedtuple('Config', [
+            'host',
+            'port',
+            'api_key']
+        )
+
+        return Config(host, port, api_key), tags
+
+    def _get_pdns_stats(self, config):
+        url = "http://{}:{}/servers/localhost/statistics".format(config.host, config.port)
+        service_check_tags = ['recursor_host:{}'.format(config.host), 'recursor_port:{}'.format(config.port)]
+        headers = {"X-API-Key": config.api_key}
+        try:
+            request = requests.get(url, headers=headers)
+            request.raise_for_status()
+            result = json.loads(request.text)
+        except Exception:
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL,
+                               tags=service_check_tags)
+            raise
```

Wouldn't we want to raise a custom error here rather than just the exception that was thrown by requests?